optuna/integration/ignite.py

```diff
@@ -0,0 +1,72 @@
+import optuna
+from optuna import type_checking
+
+if type_checking.TYPE_CHECKING:
+    from optuna.trial import Trial  # NOQA
+
+try:
+    from ignite.engine import Engine  # NOQA
+    _available = True
+except ImportError as e:
+    _import_error = e
+    # IgnitePruningHandler is disabled because pytorch-ignite is not available.
+    _available = False
+
+
+class IgnitePruningHandler(object):
+    """PyTorch Ignite handler to prune unpromising trials.
+
+    Example:
+
+        Add a pruning handler which observes validation accuracy.
+
+        .. code::
+
+                evaluator = create_supervised_evaluator(model,
+                                                        metrics={'accuracy': Accuracy()},
+                                                        device=device)
+                handler = PyTorchIgnitePruningHandler(trial, 'accuracy', trainer)
+                evaluator.add_event_handler(Events.COMPLETED, handler)
+
+                @trainer.on(Events.EPOCH_COMPLETED)
+                def log_validation_results(engine):
+                    evaluator.run(val_loader)
+
+    Args:
+        trial:
+            A :class:`~optuna.trial.Trial` corresponding to the current evaluation of the
+            objective function.
+        metric:
+            A name of metric for pruning, e.g., ``accuracy`` and ``loss``.
+        trainer:
+            A trainer engine of PyTorch Ignite. Please refer to `ignite.engine.Engine reference
+            <https://pytorch.org/ignite/engine.html#ignite.engine.Engine>`_ for further details.
+    """
+
+    def __init__(self, trial, metric, trainer):
+        # type: (Trial, str, Engine) -> None
+
+        self.trial = trial
+        self.metric = metric
+        self.trainer = trainer
+
+    def __call__(self, engine):
+        # type: (Engine) -> None
+
+        score = engine.state.metrics[self.metric]
+        self.trial.report(score, engine.state.epoch)
+        if self.trial.should_prune():
+            self.trainer.terminate()
```

Comment:
I don't know the detail of  `self.trainer.terminate()`, but I think that it's better to ensure that the method is called before aborting a trial as follows:
```python
try:
    score = engine.state.metrics[self.metric]   # may fail
    self.trial.report(score, engine.state.epoch)   # may fail
    if self.trial.should_prune():   # may fail
        message = ...
        raise optuna.structs.TrialPruned(message)
except:
    self.trainer.terminate()
    raise
```
What do you think?

Reply:
Or, I think that we might be able to simply remove the invocation of `self.trainer.terminate()`.
Could you explain why this method invocation is needed?

Reply:
As you mentioned, it should be removed from the code because it simply set [the `should_terminate` flag](https://github.com/pytorch/ignite/blob/master/ignite/engine/engine.py#L380-L384), which it will not be referred by any other part of code.
(It was introduced according to the implementation of the [early stopping handler](https://github.com/pytorch/ignite/blob/master/ignite/handlers/early_stopping.py#L63).)

Reply:
I see. Thank you for the explanation.